### PR TITLE
Switch from XVFB/pyvirtualdisplay to Firefox's native headless mode

### LINKED ISSUE #425: Cannot run demo.py
Hello, I am using MacOS Mojave Version 10.14.5 with latest Docker and trying to install OpenWPM. After installation while running demo.py, I get the following error continuously even after multiple trials of reinstall.

==============================================================

root@dafa11160f3b:/opt/OpenWPM# python demo.py
No handlers could be found for logger "pyvirtualdisplay.abstractdisplay"
BrowserManager       - INFO     - BROWSER 1: Crash in driver, restarting browser manager 
 Traceback (most recent call last):
  File "/opt/OpenWPM/automation/BrowserManager.py", line 360, in BrowserManager
    status_queue, browser_params, manager_params, crash_recovery)
  File "/opt/OpenWPM/automation/DeployBrowsers/deploy_browser.py", line 15, in deploy_browser
    manager_params, crash_recovery)
  File "/opt/OpenWPM/automation/DeployBrowsers/deploy_firefox.py", line 173, in deploy_firefox
    driver.install_addon(ext_loc, temporary=True)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/firefox/webdriver.py", line 267, in install_addon
    return self.execute("INSTALL_ADDON", payload)["value"]
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
WebDriverException: Message: No such file or directory: /opt/OpenWPM/automation/Extension/firefox/openwpm.xpi

BrowserManager       - ERROR    - BROWSER 1: Spawn unsuccessful  | Proxy Ready: False  | Profile Created: True  | Profile Tar: True  | Display: True  | Launch Attempted: True  | Browser Launched: False  | Browser Ready: False 
No handlers could be found for logger "pyvirtualdisplay.abstractdisplay"
BrowserManager       - INFO     - BROWSER 1: Crash in driver, restarting browser manager 
 Traceback (most recent call last):
  File "/opt/OpenWPM/automation/BrowserManager.py", line 360, in BrowserManager
    status_queue, browser_params, manager_params, crash_recovery)
  File "/opt/OpenWPM/automation/DeployBrowsers/deploy_browser.py", line 15, in deploy_browser
    manager_params, crash_recovery)
  File "/opt/OpenWPM/automation/DeployBrowsers/deploy_firefox.py", line 173, in deploy_firefox
    driver.install_addon(ext_loc, temporary=True)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/firefox/webdriver.py", line 267, in install_addon
    return self.execute("INSTALL_ADDON", payload)["value"]
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
WebDriverException: Message: No such file or directory: /opt/OpenWPM/automation/Extension/firefox/openwpm.xpi

BrowserManager       - ERROR    - BROWSER 1: Spawn unsuccessful  | Proxy Ready: False  | Profile Created: True  | Profile Tar: True  | Display: True  | Launch Attempted: True  | Browser Launched: False  | Browser Ready: False 
No handlers could be found for logger "pyvirtualdisplay.abstractdisplay"
BrowserManager       - INFO     - BROWSER 1: Crash in driver, restarting browser manager 
 Traceback (most recent call last):
  File "/opt/OpenWPM/automation/BrowserManager.py", line 360, in BrowserManager
    status_queue, browser_params, manager_params, crash_recovery)
  File "/opt/OpenWPM/automation/DeployBrowsers/deploy_browser.py", line 15, in deploy_browser
    manager_params, crash_recovery)
  File "/opt/OpenWPM/automation/DeployBrowsers/deploy_firefox.py", line 173, in deploy_firefox
    driver.install_addon(ext_loc, temporary=True)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/firefox/webdriver.py", line 267, in install_addon
    return self.execute("INSTALL_ADDON", payload)["value"]
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
WebDriverException: Message: No such file or directory: /opt/OpenWPM/automation/Extension/firefox/openwpm.xpi

BrowserManager       - ERROR    - BROWSER 1: Spawn unsuccessful  | Proxy Ready: False  | Profile Created: True  | Profile Tar: True  | Display: True  | Launch Attempted: True  | Browser Launched: False  | Browser Ready: False 
No handlers could be found for logger "pyvirtualdisplay.abstractdisplay"
BrowserManager       - INFO     - BROWSER 1: Crash in driver, restarting browser manager 
 Traceback (most recent call last):
  File "/opt/OpenWPM/automation/BrowserManager.py", line 360, in BrowserManager
    status_queue, browser_params, manager_params, crash_recovery)
  File "/opt/OpenWPM/automation/DeployBrowsers/deploy_browser.py", line 15, in deploy_browser
    manager_params, crash_recovery)
  File "/opt/OpenWPM/automation/DeployBrowsers/deploy_firefox.py", line 173, in deploy_firefox
    driver.install_addon(ext_loc, temporary=True)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/firefox/webdriver.py", line 267, in install_addon
    return self.execute("INSTALL_ADDON", payload)["value"]
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/usr/local/lib/python2.7/dist-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
WebDriverException: Message: No such file or directory: /opt/OpenWPM/automation/Extension/firefox/openwpm.xpi

BrowserManager       - ERROR    - BROWSER 1: Spawn unsuccessful  | Proxy Ready: False  | Profile Created: True  | Profile Tar: True  | Display: True  | Launch Attempted: True  | Browser Launched: False  | Browser Ready: False 
TaskManager          - CRITICAL - Browser spawn failure during TaskManager initialization, exiting...
Traceback (most recent call last):
  File "demo.py", line 37, in <module>
    manager = TaskManager.TaskManager(manager_params, browser_params)
  File "/opt/OpenWPM/automation/TaskManager.py", line 123, in __init__
    self._launch_browsers()
  File "/opt/OpenWPM/automation/TaskManager.py", line 161, in _launch_browsers
    self.close()
  File "/opt/OpenWPM/automation/TaskManager.py", line 571, in close
    self._shutdown_manager()
  File "/opt/OpenWPM/automation/TaskManager.py", line 266, in _shutdown_manager
    browser.shutdown_browser(during_init)
  File "/opt/OpenWPM/automation/BrowserManager.py", line 307, in shutdown_browser
    self.kill_browser_manager()
  File "/opt/OpenWPM/automation/BrowserManager.py", line 231, in kill_browser_manager
    self.crawl_id, self.browser_manager.pid, self.display_pid,
AttributeError: 'NoneType' object has no attribute 'pid'

===============================================================

### PR DESCRIPTION
**Pull request by crankycoder**
This should fix #425

This removes the dependency on xvfb and pyvirtualdisplay to run headless.

I've also added a couple other things to simplify the developer workflow a little
 * a minimal docker-compose.yml file to simplify the build to include the `openwpm` image name.  
 * Makefile with a single `devshell` target so you can run bash inside of the container to debug things when things go unexpectedly.

### CODE REVIEW

**Comment by birdsarah** [comment]
Seems you're having some problems with flash cookies. I did a little digging and headless mode should support it, but it's turned off by default (https://github.com/mozilla/geckodriver/releases?after=v0.17.0)

https://github.com/mozilla/geckodriver/issues/1103#issuecomment-352188142

You can set the `plugin.state.flash` preference to `1` to override this behaviour and enable Flash.

We may just need to add this for those tests now geckdriver is in headless mode.

**Comment by motin** [comment]
@crankycoder Could you move the Dockerfile enhancement to a separate PR for proper review? Thanks. As soon as the test_flash_cookies test is passing, this PR can be reviewed.

**Comment by englehardt** [comment]
Thanks! Not having to create and manage XVFB processes would be nice. This will require a bit of cleanup in addition to the current changes. We'll also want to remove the following:

* [passing the display `pid` and `port`](https://github.com/mozilla/OpenWPM/blob/master/automation/DeployBrowsers/deploy_firefox.py#L108), which is no longer necessary if XVFB isn't running in its own process.
* The corresponding code that handles the display info throughout BrowserManager.py and TaskManager.py. A non-exhaustive list: [here](https://github.com/mozilla/OpenWPM/blob/master/automation/BrowserManager.py#L62-L65), [here](https://github.com/mozilla/OpenWPM/blob/9f0442241909fdac273eeb444bf06c0f2beb781c/automation/BrowserManager.py#L153-L155), [here](https://github.com/mozilla/OpenWPM/blob/9f0442241909fdac273eeb444bf06c0f2beb781c/automation/BrowserManager.py#L242-L260), [here](https://github.com/mozilla/OpenWPM/blob/9a06168e95997866b64dc1b4802800f00b26fbf4/automation/TaskManager.py#L193-L216),
* Remove pyvirtualdisplay from `requirements.txt`
* Update README.md to reflect that we aren't using a virtual display and instead using Firefox's headless mode. We should also review and update the comment on macOS support.
*

**Comment by birdsarah** [comment]
I spent a couple of hours trying to get the test_flash_cookies test to pass and while I had success locally, I didn't on Travis. I spoke to @englehardt briefly about this and his response:

> if flash cookies end up causing a headache for you I'm comfortable tearing those out. Flash is click to play everywhere already and very close to EOL....actually it looks like the newest version of Chrome has already removed flash

**Comment by englehardt** [comment]
Re flash: It looks like it will require users to always [ask to activate](https://developer.mozilla.org/en-US/docs/Plugins/Roadmap#Schedule) in Firefox 69, which effectively makes Flash useless for cross-site tracking. Then, in early 2020 Flash will be removed from Firefox release entirely. So it's definitely not worth us spending any extra effort trying to support flash cookie measurement. I suggest we create a separate PR that removes all flash cookie related instrumentation, tests, and dependencies.

Similarly, we should split the headless changes out from the dockerfile / dev workflow changes.

**Comment by crankycoder** [comment]
@englehardt I think I've got rid of all the display_pid and display_port references now.

The Flash cookies test has been marked 'skip' so that the tests don't fail anymore.

File: test/test_storage_vectors.py [hunk 1]
```diff
@@ -48,6 +50,7 @@ class TestStorageVectors(OpenWPMTest):
     def get_config(self, data_dir=""):
         return self.get_test_config(data_dir)
 
+    @pytest.mark.skip('Flash cookies are no longer supported in OpenWPM')
```
Comment (birdsarah) [inline test/test_storage_vectors.py]: Can I suggest xfail. "Flash cookies do not work with headless mode."

Comment (crankycoder) [reply test/test_storage_vectors.py]: Is it that flash cookies are no longer supported - or just no longer support with headless mode?

I was under the impression it was for both head and headless cases.

That aside - I've swapped out the skip with xfail.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+version: "3"
+services:
+  openwpm:
+    image: openwpm
+    environment:
+      # Pickup the AWS secret and access key from host env
```
Comment (englehardt) [inline docker-compose.yml]: I don't think we should also bring in AWS credentials into the dockerfile. Some users may not want their credentials available to the OpenWPM container.

File: test/test_storage_vectors.py (hunk 1 above)
Comment (englehardt) [reply test/test_storage_vectors.py]: I suppose a user could still enable them if wanted. I'll make a follow-up PR that totally removes flash cookie support.

**Comment by englehardt** [comment]
LGTM. Thanks @crankycoder.

One thing to note here is that headless mode may be more identifiable than XVFB. I filed https://github.com/mozilla/OpenWPM/issues/448 summarizing these concerns, but I don't think this should be a blocker.

**Pull request merged**